File: backend/src/opensprite_backend/response_modes.py
"""Response preferences and the immutable reasoning decision used for a Run."""

from dataclasses import dataclass
from enum import StrEnum
from typing import Literal


class ResponseMode(StrEnum):
    DEFAULT = "default"
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    XHIGH = "xhigh"
    MAX = "max"
    ULTRA = "ultra"


ResponseModeValue = Literal["default", "low", "medium", "high", "xhigh", "max", "ultra"]
HistoricalResponseMode = Literal["default", "fast", "balanced", "deep", "low", "medium", "high", "xhigh", "max", "ultra"]
RESPONSE_MODES = tuple(mode.value for mode in ResponseMode)
LEGACY_RESPONSE_MODES = {"fast": "low", "balanced": "medium", "deep": "high"}
HISTORICAL_RESPONSE_MODES = (*LEGACY_RESPONSE_MODES, *RESPONSE_MODES)
NATIVE_EFFORTS = ("none", "minimal", "low", "medium", "high", "xhigh", "max")


def migrate_response_mode(value: object) -> str:
    """Storage migrations preserve default and translate retired preference names."""
    if not isinstance(value, str) or value not in HISTORICAL_RESPONSE_MODES:
        raise ValueError("invalid response mode")
    return LEGACY_RESPONSE_MODES.get(value, value)


@dataclass(frozen=True, slots=True)
class ReasoningResolution:
    requested: HistoricalResponseMode
    effective: str | None
    status: Literal["exact", "fallback", "provider_default", "unknown"]

    def __post_init__(self) -> None:
        if self.requested not in HISTORICAL_RESPONSE_MODES:
            raise ValueError("invalid requested response mode")
        if self.status in {"exact", "fallback"}:
            if self.effective not in NATIVE_EFFORTS:
                raise ValueError("invalid effective reasoning effort")
        elif self.status not in {"provider_default", "unknown"} or self.effective is not None:
            raise ValueError("invalid reasoning resolution")


def resolve_response_mode(mode: HistoricalResponseMode, supported: tuple[str, ...] | None) -> ReasoningResolution:
    if mode == "default":
        return ReasoningResolution(mode, None, "provider_default")
    requested = migrate_response_mode(mode)
    if supported is None:
        return ReasoningResolution(mode, None, "unknown")
    available = tuple(value for value in NATIVE_EFFORTS if value in supported)
    if not available:
        return ReasoningResolution(mode, None, "provider_default")
    if requested in available:
        return ReasoningResolution(mode, requested, "exact")
    return ReasoningResolution(mode, available[-1], "fallback")

File: frontend/src/api/responseModes.ts
import { apiFetch } from "./http";
import type { ProviderId } from "./providerConnections";

export const responseModes = ["default", "low", "medium", "high", "xhigh", "max", "ultra"] as const;
export type ResponseMode = typeof responseModes[number];
export const historicalResponseModes = ["fast", "balanced", "deep", ...responseModes] as const;
export type HistoricalResponseMode = typeof historicalResponseModes[number];
export const nativeEfforts = ["none", "minimal", "low", "medium", "high", "xhigh", "max"] as const;
export type ReasoningResolution = {
  requested: HistoricalResponseMode;
  effective: typeof nativeEfforts[number] | null;
  status: "exact" | "fallback" | "provider_default" | "unknown";
};

export function isReasoningResolution(value: unknown): value is ReasoningResolution {
  if (typeof value !== "object" || value === null || Array.isArray(value)) return false;
  const item = value as Record<string, unknown>;
  if (Object.keys(item).length !== 3 || !historicalResponseModes.includes(item.requested as HistoricalResponseMode)) return false;
  if (item.status === "exact" || item.status === "fallback") return nativeEfforts.includes(item.effective as typeof nativeEfforts[number]);
  return (item.status === "provider_default" || item.status === "unknown") && item.effective === null;
}

export async function getResponseModeResolution(providerId: ProviderId, modelId: string, responseMode: ResponseMode, signal: AbortSignal): Promise<ReasoningResolution> {
  const query = new URLSearchParams({ providerId, modelId, responseMode });
  const response = await apiFetch(`/api/settings/ai/response-mode?${query}`, { signal });
  if (!response.ok) throw new Error("response_mode_unavailable");
  const value: unknown = await response.json();
  if (!isReasoningResolution(value) || value.requested !== responseMode) throw new Error("invalid_response_mode_resolution");
  return value;
}
